engine/src/flutter/impeller/geometry/rational.cc

```diff
@@ -0,0 +1,58 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/impeller/geometry/rational.h"
+
+#include <cmath>
+#include <cstdlib>
+#include <numeric>
+
+namespace impeller {
+namespace {
+uint32_t AbsToUnsigned(int32_t x) {
+  return static_cast<uint32_t>(std::abs(x));
+}
+}  // namespace
+
+bool Rational::operator==(const Rational& that) const {
+  if (den_ == that.den_) {
+    return num_ == that.num_;
+  } else if ((num_ >= 0) != (that.num_ >= 0)) {
+    return false;
+  } else {
+    return AbsToUnsigned(num_) * that.den_ == AbsToUnsigned(that.num_) * den_;
+  }
+}
+
+bool Rational::operator!=(const Rational& that) const {
+  return !(*this == that);
+}
+
+bool Rational::operator<(const Rational& that) const {
+  if (den_ == that.den_) {
+    return num_ < that.num_;
+  } else if ((num_ >= 0) != (that.num_ >= 0)) {
+    return num_ < that.num_;
+  } else {
+    return AbsToUnsigned(num_) * that.den_ < AbsToUnsigned(that.num_) * den_;
+  }
+}
+
+uint64_t Rational::GetHash() const {
+  if (num_ == 0) {
+    return 0;
+  }
+  uint64_t gcd = std::gcd(num_, den_);
+  return ((num_ / gcd) << 32) | (den_ / gcd);
```

Comment:
So the goal here is to compute a hash that is == when the values are equal, basically making sure that 1/5 and 2/10 have the same hash?

Reply:
Exactly, we had a few other options:
1) Simplify the fractions in the constructor so that if you ask for Rational(2, 10) you get Rational(1, 5), but I tried to optimize the math for fractions with the same denominator.
1) Make the denominator part of the type, so `Rational(1,5)` becomes `Rational<5>(1)`. That way we could have made sure everything has the same denominator.  It only occurred to me later, it is a bit awkward though.

Reply:
I'm slightly concerned about the performance of computing the gcd (nlogn ?) for the hash, given that we end up hashing each glyph each frame. the subpixel hash already shows up quite hot in the flame graphs. this can be verified on an app with a moderate amount of text, like the old flutter gallery.

I'd lean towards an approach where we computed the float and then hashed that.

Reply:
I had the same concern, but gcd isn't actually that expensive.  Our denominator will typically be 200, so the worst case scenario is Rational(199,200).  The euclidean method loops 3 times in order to figure that out which is like 12 operations.

I don't think we want to compute the float since you'll be throwing away precision.  We have the opportunity to have a perfect hash function without gcd in it in the 2 approaches I listed above.

Reply:
Let's land it like this and look at the numbers afterwards.  I say if we notice any change we should switch to approach 2 since it is the fastest.